internal/lifecycle/readme.go

```diff
@@ -0,0 +1,206 @@
+// that code block.
+func codeBlocks(scanner *bufio.Scanner) ([][]string, error) {
+	var blocks [][]string
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.Contains(line, codeTag) {
+			if s := scanner.Scan(); !s {
+				if err := scanner.Err(); err != nil && !s {
+					return nil, fmt.Errorf("[lifecycle.codeBlocks] bufio.Scanner: %w", err)
+				}
+				return nil, fmt.Errorf("[lifecycle.codeBlocks]: unexpected EOF; file ended immediately " +
+					"after code tag")
+			}
+
+			startCodeBlockLine := scanner.Text()
+			m := mdCodeFenceStartRegexp.MatchString(startCodeBlockLine)
+			if !m {
+				return nil, fmt.Errorf("[lifecycle.codeBlocks]: expecting start of code block immediately " +
+					"after code tag")
+			}
+
+			c := strings.Count(startCodeBlockLine, "`")
+			mdCodeFenceEndRegexp := regexp.MustCompile(fmt.Sprintf("^\\w*`{%d,}\\w*$", c))
+
+			var block []string
+			var blockClosed bool
+			for scanner.Scan() {
+				line = strings.TrimSpace(scanner.Text())
+				if mdCodeFenceEndRegexp.MatchString(line) {
+					blockClosed = true
+					break
+				}
+
+				block = append(block, line)
+			}
+
+			if err := scanner.Err(); err != nil {
+				return nil, fmt.Errorf("[lifecycle.codeBlocks] bufio.Scanner: %w", err)
+			}
+
+			if !blockClosed {
+				return nil, fmt.Errorf("[lifecycle.codeBlocks]: unexpected EOF; code block not closed")
+			}
+
+			blocks = append(blocks, block)
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("[lifecycle.codeBlocks] bufio.Scanner: %w", err)
+	}
+
+	return blocks, nil
+}
+
+// codeBlocksTolifecycle takes a slice of Markdown code blocks and extracts the terminal commands contained within them
+// to form a Lifecycle. It handles the expansion of environment variables and line continuations. It also detects
+// Cloud Run service names Google Container Registry container image URLs and replaces them with the ones provided.
+func codeBlocksTolifecycle(codeBlocks [][]string, serviceName, gcrURL string) (Lifecycle, error) {
+	var lifecycle Lifecycle
+
+	for _, block := range codeBlocks {
+		for i := 0; i < len(block); i++ {
+			line := block[i]
+			if line == "" {
+				continue
+			}
+
+			// If there is a backslash at the end of the line, this is a multiline command. Keep scanning to get
+			// entire command.
+			for line[len(line)-1] == bashLineContChar {
+				line = line[:len(line)-1]
+
+				i++
+				if i >= len(block) {
+					return nil, fmt.Errorf("[lifecycle.codeBlocksTolifecycle]: unexpected end of code block; " +
+						"expecting command line continuation")
+				}
+
+				l := block[i]
+				if l == "" {
+					break
+				}
+
+				line = line + l
+			}
+
+			line = os.ExpandEnv(line)
+			line = gcrURLRegexp.ReplaceAllString(line, gcrURL)
+			line = replaceServiceName(line, serviceName)
+			sp := strings.Split(line, " ")
+
+			var cmd *exec.Cmd
+			if strings.Contains(line, "gcloud") {
```

Comment:
Should this be "contains" or `sp[0] == "gcloud"`?

Reply:
Resolved, thank you